### ISSUE TITLE
Updating and accessing @State properties

### ISSUE THREAD

**Issue by shayna34** [nalexn/ViewInspector#5]
I have the following view, in which an optional `Text` view is toggled in and out of visibility when the button is tapped:
```
struct MyView: View {

    @State var showErrorMessage = false

    var body: some View {
        VStack {
            Button(action: {
                self.showErrorMessage.toggle()
                print("showErrorMessage = \(self.showErrorMessage)")
            }) {
                Text("Click me.")
            }

            if showErrorMessage {
                Text("ERROR")
                    .foregroundColor(Color.red)
            }
        }
    }
}
```
I am trying to test the optional view with the following:
```
it("should show an error message when the button is pressed") {
        do {
            let view = MyView()
            let button = try view.inspect().vStack().button(0)

            try button.tap()

            let errorMessage = try view.inspect().vStack().text(1).string()

            expect(view.showErrorMessage).to(beTrue())
            expect(errorMessage).to(equal("ERROR"))
        } catch {
            fail("failure - \(error)")
        }
 }
```
The test fails on both assertions, as the `showErrorMessage` property remains false and the optional `Text` view remains hidden. Additionally, `showErrorMessage = false` is printed to the console when `try button.tap()` is called.

When I run the code in the UI simulator, it works as expected; the `Text` view is toggled in and out of visibility and the correct value of `showErrorMessage` is printed to the console.

Is there something I'm missing that's causing the different behavior in the test versus the UI simulator?

**Comment by nalexn** [comment]
SwiftUI handles `@State` mutation in a quite peculiar way: `@State` works as a proxy forwarding the updates to dedicated value storage in the SwiftUI engine, while the struct itself remains unchanged. I'm exploring the options to work around this.
This issue is not apparent for `@ObservedObject` or `@EnvironmentObject`.

**Comment by nalexn** [comment]
@shayna34 good news: I've found the way to track the @State changes! I'll update the readme and tweak the framework for convenience, but conceptually here is the workaround:

Tweaks for the client view:

```swift
struct MyView: View {

    @State var showErrorMessage = false
    var didAppear: ((Self) -> Void)?

    var body: some View {
        ...
        .onAppear { self.didAppear?(self) }
    }
}
```

The code for testing:
```swift
final class StateMutationTests: XCTestCase {
    var window: UIWindow!

    func testState() {
        let exp = XCTestExpectation(description: "didAppear")
        var view = MyView()
        view.didAppear = { view in
            do {
                print(">> \(view. showErrorMessage)") // false
                try view.inspect().vStack().button(0).tap()
                print(">> \(view. showErrorMessage)") // true
                exp.fulfill()
            } catch _ { }
        }
        let vc = UIHostingController(rootView: view)
        window = UIWindow(frame: UIScreen.main.bounds)
        window.rootViewController = vc
        window.makeKeyAndVisible()
        wait(for: [exp], timeout: 0.1)
    }

```
I'll update the ViewInspector to provide a nice API which will hide all this bulky setup. Thanks for raising the issue!

**Comment by shayna34** [comment]
Oh, awesome! Works like a charm :) I'll keep an eye out for the ViewInspector update. Thank you so much for your help!

**Comment by nalexn** [comment]
It turned out that this hack allowed me to overcome the issue with `@Environment` injection! I'm preparing a major update with natural environment injection support, without the need to refactor the body. All thanks to you @shayna34 !

**Comment by nalexn** [comment]
@shayna34  The update is out! Please let me know if all works as expected

**Comment by shayna34** [comment]
@nalexn Apologize for the delay; I was able to get it working as expected. Thank you again for all of your help and for building such a useful library!

**Comment by nalexn** [comment]
That's great! Let me know if you hit another roadblock - I'm always ready to help!

**Comment by davdroman** [comment]
Sorry for commenting on a closed issue, but I wanted to personally thank you for this effort. I thought I hit a roadblock with this and here comes the Swift open source community to prove me wrong! Thank you so much.

**Comment by nalexn** [comment]
@davdroman you're very much welcome! And thank you for sponsoring! 😁
